### PR TITLE
check operand StatefulSet is ready for realm import Job to run

### LINKED ISSUE #24526: Keycloak operator 22.0.z created RealmImport Job withouth checking Keycloak operand health first
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

operator

### Describe the bug

Using Keycloak 22.0.5 build, when the `KeycloakRealmImport` CR is created soon after the primary `Keycloak` CR (e.g. in the GitOps pipeline), the real import Job `<realm-import-cr-name>` is started unconditionally and the first few Pods created (up to 6 backoff limit) might be in the failed state, till Keycloak operand fully is initialized.

This ends up with some bad initial experience where the failed Pods are found on the cluster.

While I do understand that perhaps the best practice is to wait for the primary `Keycloak` CR would be reporting Ready in `.status`, before any other Keycloak CRs are created, but it's not something typically possible / used in the GitOps pipelines.

The proposed solution is to have Keycloak Operator checking the referenced Keycloak CR status, before submtting the real import Job.

```yaml
apiVersion: k8s.keycloak.org/v2alpha1
kind: KeycloakRealmImport
metadata:
  name: cs-cloudpak-realm
spec:
  keycloakCRName: cs-keycloak-test1 <--- here, the Keycloak CR cs-keycloak-test1 is created, but its status is not Ready
  realm:
    displayName: cloudpak
    enabled: true
    id: cloudpak
    realm: cloudpak
```

So I guess there is a basic check for existence of operand `StatefulSet` at https://github.com/keycloak/keycloak/blob/main/operator/src/main/java/org/keycloak/operator/controllers/KeycloakRealmImportController.java#L116C23-L116C23 , to check whether operand StatefulSet exists, but it's not checked whether it's running, before the import Job is submitted.

### Version

22.0.5

### Expected behavior

No failed Pods when KeycloakRealmImport CR is created immediatelly after Keycloak CR (before Keycloak is ready)

### Actual behavior

Failed Pods when KeycloakRealmImport CR is created immediatelly after Keycloak CR (before Keycloak is ready)

### How to Reproduce?

Create KeycloakRealmImport CR immediatelly after Keycloak CR

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by pgodowski**
Check whether the Keycloak operand `StatefulSet` has at least single replica ready before submitting the RealmImport Job.

Closes https://github.com/keycloak/keycloak/issues/24526

cc @shawkins

### CODE REVIEW

**Review by shawkins — COMMENTED** [review]
Thank you @pgodowski this seems good - polling is simpler than getting the Keycloak CR events.  

It should have a test though.

Also I know there always should be a status added when a statefulset is created, but the code is checking for null in one place, but not the other - it should at least be consistent.

**Comment by pgodowski** [comment]
Thanks for the review.

> Also I know there always should be a status added when a statefulset is created, but the code is checking for null in one place, but not the other - it should at least be consistent.

Fixed this part.

> It should have a test though.

Will work on adding a test case in `RealmImportTest.java`

**Review by vmuzikar — COMMENTED** [review]
@pgodowski Thanks for the PR!

I believe this affects `main` too where we [don't check](https://github.com/keycloak/keycloak/blob/b343f87c606a0478a0b65ac585dbb44c7dd541d7/operator/src/main/java/org/keycloak/operator/controllers/KeycloakRealmImportController.java#L72-L73) number of ready instances as well. We should create a parallel fix for it too to consider this fully fixed.

**Comment by pgodowski** [comment]
Added unit test to check for the failed realm import when Keycloak operand is not available.

**Comment by shawkins** [comment]
@pgodowski looks good.  Please rebase to a single commit and add your DCO signature - Keycloak has recently adopted a requirement.

**Review by vmuzikar — APPROVED** [review]
LGTM (once comment from @shawkins is resolved).

**Comment by pgodowski** [comment]
> Please rebase to a single commit and add your DCO signature - Keycloak has recently adopted a requirement.

@shawkins done - thanks!

**Review by shawkins — APPROVED** [review]
LGTM.  Once small nit, could there be a single crSelector.get() rather than 3 calls per polling.

**Comment by pgodowski** [comment]
Do you want me to add any changes here?

**Comment by shawkins** [comment]
> Do you want me to add any changes here?

It's optional.  It can be cleaned up later if needed.

File: operator/src/test/java/org/keycloak/operator/testsuite/integration/RealmImportTest.java [hunk 1]
```diff
@@ -193,4 +194,33 @@ public void testNotWorkingRealmImport() {
...
     }
 
+    @Test
+    public void testFailedRealmImportWhenKeycloakNotAvailable() {
+        // Arrange
+        Keycloak kc = getDefaultKeycloakDeployment();
+        kc.getSpec().setInstances(0);
+
+        // don't wait for Keycloak being available, since it has no instances
+        deployKeycloak(k8sclient, getDefaultKeycloakDeployment(), false);
```
Comment (shawkins) [inline operator/src/test/java/org/keycloak/operator/testsuite/integration/RealmImportTest.java:204]: This should reference the kc instance created above.

**Comment by shawkins** [comment]
@pgodowski I hope you don't mind, I took the liberty of opening a forward port: #24720

There's still another small change needed here.

File: operator/src/test/java/org/keycloak/operator/testsuite/integration/RealmImportTest.java (hunk 1 above)
Comment (pgodowski) [reply operator/src/test/java/org/keycloak/operator/testsuite/integration/RealmImportTest.java]: OMG, good catch @shawkins - fixed!

**Comment by pgodowski** [comment]
Thanks for the forward port - let me know you need any inputs on that. Updated the test case, yet, for some reason I cannot now squash commits, hope it is fine?

**Review by shawkins — APPROVED** [review]
LGTM, don't worry about squashing, it will be done as part of the merge.

**Review by vmuzikar — APPROVED** [review]

**Comment by vmuzikar** [comment]
Enabled auto-merge but 22 is frozen until 22.0.6 is released.

**Comment by pgodowski** [comment]
> As an FYI (because I'm expecting this one in 22.0.7 anyway), this is actually more dangerous than we thought.
> 


Are you suggesting we shall check for other thing in the proposed fix than the operand StatefulSet, or it's just your notice saying that the impact is larger than we initially though?

**Comment by Jamstah** [comment]
> > As an FYI (because I'm expecting this one in 22.0.7 anyway), this is actually more dangerous than we thought.
> 
> Are you suggesting we shall check for other thing in the proposed fix than the operand StatefulSet, or it's just your notice saying that the impact is larger than we initially though?

Just that the impact is more than just a UX snag - it can leave an installation in a broken state.

I don't think we need more than a single pod ready check.

**Pull request merged**